File: CPP/No503.cc
/**
 * Created by Xiaozhong on 2020/10/5.
 * Copyright (c) 2020/10/5 Xiaozhong. All rights reserved.
 */
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int> &nums) {
        vector<int> ans(nums.size());
        stack<int> stk;
        for (int i = 2 * nums.size() - 1; i >= 0; --i) {
            while (!stk.empty() && nums[stk.top()] <= nums[i % nums.size()]) stk.pop();
            ans[i % nums.size()] = stk.empty() ? -1 : nums[stk.top()];
            stk.push(i % nums.size());
        }
        return ans;
    }
};

File: Java/src/No503.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Arrays;
import java.util.Stack;

public class No503 {
    public int[] nextGreaterElements(int[] nums) {
        int size = nums.length;
        int[] ret = new int[size];
        Arrays.fill(ret, -1);
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < size * 2 - 1; ++i) {
            while (!stack.empty() && (nums[stack.peek()] < nums[i % size]))
                ret[stack.pop()] = nums[i % size];
            stack.push(i % size);
        }
        return ret;
    }

    public static void main(String[] args) {
        No503 test = new No503();
        int[] nums = new int[]{1, 2, 3, 4, 3};
        System.out.println(Arrays.toString(test.nextGreaterElements(nums)));
    }
}
